### ISSUE TITLE
Video decoder threading issue

### ISSUE THREAD

**Issue by SteveKChiu** [wang-bin/QtAV#796]
It seems some codec is not thread safe (e.g. ffmpeg android MediaCodec), that it requires alloc/free in the same thread, otherwise something unexpected may happen.

I have observed huge leak of local reference of java objects while using MediaCodec decoder. After hunting for the issue, i find the ffmpeg implementation is using JNI AttachCurrentThread/DetachCurrentThread in the alloc/free function. As result, you must call all the codec related functions in the same thread.

**Comment by SteveKChiu** [comment]
That includes av_frame_alloc and av_frame_free, because some codec related context may reside in that structure.

**Comment by wang-bin** [comment]
I know it's bad and even not correct to create/use/destroy a decoder in different threads. I'm not sure whether it affects mediacodec. 

> I have observed huge leak of local reference of java objects while using MediaCodec decoder
In ffmpeg or your code?

> That includes av_frame_alloc and av_frame_free, because some codec related context may reside in that structure.
Yes, for hw decoders.

**Comment by SteveKChiu** [comment]
I am using QtAV QML MediaPlayer with property:

````
MediaPlayer {
...
videoCodecPriority: ["MediaCodec", "FFmpeg"]
...
}
````

I have a queue of MediaPlayer that get created and destroyed (for seamless playback),

The video plays with threading warning, something like:

````
W art     : Native thread exiting without having called DetachCurrentThread (maybe it's going to use a pthread_key_create destructor?): Thread[20,tid=11993,Native,Thread*=0x9967b900,peer=0x12c4a220,"Thread-879"]
````

And then the app eventually crashed, with local references table overflow.

I guess that is because I create/destroy so many MediaPlayers, so the leak add up very quickly and crash the app.

````
F art     : art/runtime/indirect_reference_table.cc:115] JNI ERROR (app bug): local reference table overflow (max=512)
````

Calling av_frame_free in different thread  may cause problem because ffmpeg's MediaCodec implementation call AttachCurrentThread  on alloc and call DetachCurrentThread on free, any calls in between should be in the same thread.

**Comment by wang-bin** [comment]
> W art     : Native thread exiting without having called DetachCurrentThread (maybe it's going to use a pthread_key_create destructor?): Thread[20,tid=11993,Native,Thread*=0x9967b900,peer=0x12c4a220,"Thread-879"]

It's ok. The warning always appears when using jni env with thread local.

> F art     : art/runtime/indirect_reference_table.cc:115] JNI ERROR (app bug): local reference table overflow (max=512)

I think it's a ffmpeg bug

**Comment by wang-bin** [comment]
I confirmed the leak is a ffmpeg bug. fixed in ffmpeg 3.3
